lambdas/data-migration2/src/granulesAndFiles.ts

```diff
@@ -43,15 +45,24 @@ export const migrateGranuleRecord = async (
   Manager.recordIsValid(record, schemas.granule);
   const { name, version } = deconstructCollectionId(record.collectionId);
   const collectionPgModel = new CollectionPgModel();
+  const executionPgModel = new ExecutionPgModel();
   const granulePgModel = new GranulePgModel();
 
   const collectionCumulusId = await collectionPgModel.getRecordCumulusId(
     knex,
     { name, version }
   );
 
-  let existingRecord;
+  // TODO: Is this safe? `execution` is a required field from Dynamo now,
+  // but maybe not for older records
+  const executionCumulusId = await executionPgModel.getRecordCumulusId(
```

Comment:
That's a reasonable concern.   It's probably worth trapping the failed return and doing the original upsert in that instance.

Reply:
I also notice in the 'nullable' unit test we're not testing that case currently.

Reply:
Yeah, it wasn't included in the `nullable` test because it is a required field as of the current schema. And that goes back 4 years. 

That said, an execution is now effectively required by the `upsert` query for granules, because the query builder will fail if one is not provided:

https://github.com/nasa/cumulus/pull/2091/files?file-filters%5B%5D=.ts#diff-6307984413f5b3418c109453a08f069ff55f4c3b68430454c19e96c714719f8cR36

So I'm not sure what we do in the case where `record.execution` doesn't exist? Fail the migration of the granule record and force them to associate an execution? I think this is an extreme edge case, but still

Reply:
>  required field as of the current schema. 

Current being the API?   Yeah fair -- I did a quick browse through the old api migrations and it doesn't look like we did anything to add a default value or something weird like that.... so it's *possible*, even probable nobody has a long-running instance without an execution. 

> So I'm not sure what we do in the case where record.execution doesn't exist? 

I think if we think it's in the realm of *possible* there's a Core database that old we should: 

1) Allow the write given granule-execution links aren't required by the db schema (yikes) and consider the implications for downstream reads (e.g. are we failing badly anywhere a "get granule" assumes there should be an execution in the business logic) 

or

2)  Populate a stub execution record to associate any records that don't have one  on migration

If we *don't* think it's in the realm of possible in a production environment, then failing seems fine.

Reply:
I don't love either option but I'm going with option 1. If we're not going to require users to have an execution for a granule or accept that we should fail a record migration without an execution, then this feels preferable to creating a fake execution record.

7e8229d819f999d3d5f54ab021e02a6cc7f4cec1
de2ed304d9e5e117421aeb2d316f52f21ff955dc
e0742134bfa38f07f604b9c3e261435ae37fb7ff

I don't like that the result is that `granulePgModel.upsert()` and `upsertGranuleWithExecutionJoinRecord` now treat the execution as optional even when in reality we want it to be required for all forward writes. But we can't have it both ways.

Reply:
> I don't like that the result is that granulePgModel.upsert() and upsertGranuleWithExecutionJoinRecord now treat the execution as optional even when in reality we want it to be required for all forward writes. But we can't have it both ways.

🤔  Could we have it both ways if just explicitly write `upsertWithoutExecution`, and use it only for the migration?

Reply:
Oh man. I pretty much have this implemented by just doing `granulePgModel.upsert()` and `await granulesExecutionsPgModel.upsert(` conditionally when the execution is found in PG, **however** when unit testing that I noticed that we validate the record according to the Dynamo schema before attempting to migrate it:

`Manager.recordIsValid(record, schemas.granule);`

And since `execution` is required in the Dynamo schema, this means that the record migration isn't even attempted if that property is missing. So this concern is not valid unless we're going to change that aspect of how the migration works